### PR TITLE
python312Packages.tinygrad: 0.9.0 -> 0.9.2

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Description of changes

Changelog: https://github.com/tinygrad/tinygrad/releases/tag/v0.9.2

cc @SomeoneSerge @wozeparrot

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 342231` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.tinygrad</li>
    <li>python311Packages.tinygrad.dist</li>
    <li>python312Packages.tinygrad</li>
    <li>python312Packages.tinygrad.dist</li>
  </ul>
</details>

**Comment by wozeparrot** [comment]
Is this comment and corresponding patch still needed as rocm 6 is in nixpkgs now? https://github.com/GaetanLepage/nixpkgs/blob/fcb53e45ad5edcd50b97a0c4365f8596a20a8dec/pkgs/development/python-modules/tinygrad/default.nix#L63-L71

**Comment by GaetanLepage** [comment]
> Is this comment and corresponding patch still needed as rocm 6 is in nixpkgs now? https://github.com/GaetanLepage/nixpkgs/blob/fcb53e45ad5edcd50b97a0c4365f8596a20a8dec/pkgs/development/python-modules/tinygrad/default.nix#L63-L71

You are definitely right.
Unfortunately, `torch` is currently marked as broken when `rocmSupport` is enabled, so I had to manually remove the problematic test dependencies for the derivation to evaluate.

I adjusted the patches as needed but got two weird failures:
- 1) `extra/hip_gpu_driver/test_kfd_2.py` still contains a reference to `[/usr/include/linux/kfd_ioctl.h](https://github.com/tinygrad/tinygrad/pull/5917)`. I have tried to inject the path `${linuxHeaders}/include/kfd_ioct.h` which points to [this file](https://github.com/torvalds/linux/blob/v6.9/include/uapi/linux/kfd_ioctl.h) on my system. This leads to:
```
AttributeError: module 'tinygrad.runtime.autogen.kfd' has no attribute 'struct_kfd_ioctl_criu_args'. Did you mean: 'struct_kfd_ioctl_svm_args'?
```
I have noticed that https://github.com/tinygrad/tinygrad/pull/5917 vendored a compatible version of this file but an the system path `/usr/include/linux/kfd_ioctl.h` has been left out (by mistake ?) in [`test_kfd_2.py`](https://github.com/tinygrad/tinygrad/blob/v0.9.2/extra/hip_gpu_driver/test_kfd_2.py#L31).
I have patched this line to use the vendored copy and it seems to work.

- 2)  `extra/hip_gpu_driver/test_sdma_fun.py` fails at collection with:
```
    from tinygrad.runtime.ops_amd import AMDDevice, kio, sdma_pkts, libc
E   ImportError: cannot import name 'sdma_pkts' from 'tinygrad.runtime.ops_amd' (/build/source/tinygrad/runtime/ops_amd.py)
```

**Comment by GaetanLepage** [comment]
According to upstream, files under `extra/` are not considered as conventional tests and should not be included in the test suite.
I then dropped the related patches and references to `extra` in the derivation.

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 342231` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.tinygrad</li>
    <li>python311Packages.tinygrad.dist</li>
    <li>python312Packages.tinygrad</li>
    <li>python312Packages.tinygrad.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 342231` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.tinygrad</li>
    <li>python311Packages.tinygrad.dist</li>
    <li>python312Packages.tinygrad</li>
    <li>python312Packages.tinygrad.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/tinygrad/default.nix
Comment (GaetanLepage) [reply pkgs/development/python-modules/tinygrad/default.nix]: Definitely, this is a mistake. I thought that it was about the `hipGetDevicePropertiesR0600` too.
I have just put it back.

**Comment by wozeparrot** [comment]
The `nvrtc` patch should still be required? `tinygrad/runtime/autogen/cuda.py` was just split into `tinygrad/runtime/autogen/cuda.py` and `tinygrad/runtime/autogen/nvrtc.py`: https://github.com/tinygrad/tinygrad/blob/v0.9.2/tinygrad/runtime/autogen/nvrtc.py

**Comment by GaetanLepage** [comment]
> The `nvrtc` patch should still be required? `tinygrad/runtime/autogen/cuda.py` was just split into `tinygrad/runtime/autogen/cuda.py` and `tinygrad/runtime/autogen/nvrtc.py`: https://github.com/tinygrad/tinygrad/blob/v0.9.2/tinygrad/runtime/autogen/nvrtc.py

Good catch, done !

**Comment by GaetanLepage** [comment]
Everything should be good now.
It builds fine both with and without CUDA.

I will wait for @SomeoneSerge's review before merging.

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 342231` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.tinygrad</li>
    <li>python311Packages.tinygrad.dist</li>
    <li>python312Packages.tinygrad</li>
    <li>python312Packages.tinygrad.dist</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 342231` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages built:</summary>
  <ul>
    <li>python311Packages.tinygrad</li>
    <li>python311Packages.tinygrad.dist</li>
    <li>python312Packages.tinygrad</li>
    <li>python312Packages.tinygrad.dist</li>
  </ul>
</details>

File: pkgs/development/python-modules/tinygrad/fix-dlopen-cuda.patch
Comment (SomeoneSerge) [inline pkgs/development/python-modules/tinygrad/fix-dlopen-cuda.patch:10]: This just unfolds the `_try_dlopen` helper?

Comment (GaetanLepage) [reply pkgs/development/python-modules/tinygrad/fix-dlopen-cuda.patch:10]: Yes, I removed the function because there are no more 2 libs now.

File: pkgs/development/python-modules/tinygrad/default.nix
Comment (SomeoneSerge) [inline pkgs/development/python-modules/tinygrad/default.nix:69]: Was this also related to rocm6 or why are we removing this?

Comment (GaetanLepage) [reply pkgs/development/python-modules/tinygrad/default.nix:69]: This file is within `extra` and now we don't really care about `extra` anymore, so I removed it.
I could bring it back though.

Comment (SomeoneSerge) [reply pkgs/development/python-modules/tinygrad/default.nix:69]: Idk what extra was and why we cared before but I trust you jaja

**Review by SomeoneSerge — APPROVED** [review]
I'm not sure anyone tested rocm, but if anything breaks we can follow up in another pr...

File: pkgs/development/python-modules/tinygrad/fix-dlopen-cuda.patch
Comment (GaetanLepage) [reply pkgs/development/python-modules/tinygrad/fix-dlopen-cuda.patch:10]: Would you be interested in having that @wozeparrot ? If so I can make a PR.

**Comment by GaetanLepage** [comment]
> I'm not sure anyone tested rocm, but if anything breaks we can follow up in another pr...

As I said, the package currently doesn't evaluate because `torchWithRocm` is marked as broken.
However, I haven't noticed any rocm-related failure when removing torch from `nativeCheckInputs`

**Comment by SomeoneSerge** [comment]
> As I said, the package currently doesn't evaluate because torchWithRocm is marked as broken.
However, I haven't noticed any rocm-related failure when removing torch from nativeCheckInputs

Well that rather means we can't test what effects any rocm changes might have, so maybe we should've abstained from rocm-related refactorings, but I guess we figure it out when we unbreak torch and start unbreaking tinygrad-rocm

File: pkgs/development/python-modules/tinygrad/fix-dlopen-cuda.patch
Comment (wozeparrot) [reply pkgs/development/python-modules/tinygrad/fix-dlopen-cuda.patch:10]: Yea, should be fine to pr, would need to change in the `autogen_stubs.sh` file. We have a `try_dlopen` thing in autogen but its only used for amd_comgr right now: https://github.com/tinygrad/tinygrad/blob/5a7cb8d5a55bb4f5dc2dcf6dd56cd6bfd3b8f76c/autogen_stubs.sh#L25

**Pull request merged**
